File: services/foodService.go
package services

import (
	"net/http"

	"github.com/gin-gonic/gin"
	"github.com/google/uuid"

	"kerokume-go/config"
	"kerokume-go/repos"
	"kerokume-go/schemas"
	"kerokume-go/schemas/contracts"
	"kerokume-go/schemas/validates"
	"kerokume-go/utils"
)

func FoodServiceCreate(ctx *gin.Context) {
	logger := config.NewLogger("FOOD CREATE HANDLER")
	var dto contracts.FoodRequest
	if err := ctx.BindJSON(&dto); err != nil {
		utils.SendError(ctx, 400, err.Error())
		return
	}

	if err := validates.ValidateFoodRequest(&dto); err != nil {
		logger.Errf("validation error: %v", err)
		utils.SendError(ctx, 400, err.Error())
		return
	}

	_, err := repos.FindUniqueMenu(dto.MenuId, ctx)
	if err != nil {
		utils.SendError(ctx, http.StatusUnauthorized, "Not exist a Menu with this id")
		return
	}

	if dto.FoodCategory != "COMIDA" && dto.FoodCategory != "BEBIDA" {
		utils.SendError(ctx, http.StatusUnauthorized, "Not is permited this food category")
		return
	}

	food := schemas.Food{
		Name:         dto.Name,
		Description:  dto.Description,
		Price:        dto.Price,
		PathImg:      "",
		FoodCategory: dto.FoodCategory,
		MenuId:       dto.MenuId,
	}

	repos.SaveFood(&food, ctx)
	utils.SendSuccess(ctx, "create-food")
}

func FoodServiceGetAll(ctx *gin.Context) {
	foods, err := repos.FindAllFood(ctx)
	if err != nil {
		utils.SendError(ctx, http.StatusInternalServerError, "error finding all foods")
		return
	}
	data := make([]interface{}, len(foods))
	for i := range foods {
		data[i] = contracts.FoodResponse{
			Id: foods[i].ID,
			Name: foods[i].Name,
			Description: foods[i].Description,
			Price: foods[i].Price,
			PathImg: foods[i].PathImg,
			FoodCategory: foods[i].FoodCategory,
			IsAvailable: foods[i].IsAvailable,
		}
	}
	utils.SendSuccessArray(ctx, "find-all-foods", data)
}

func FoodServiceGetByMenuID(ctx *gin.Context) {
	menuIdStr := ctx.Param("id")
	menuId, err := uuid.Parse(menuIdStr)
	if err != nil {
		utils.SendError(ctx, http.StatusBadRequest, "invalid menu id")
		return
	}
	foods, err := repos.FindAllFoodByMenuId(menuId, ctx)
	if err != nil {
		utils.SendError(ctx, http.StatusInternalServerError, "error finding all foods by menu id")
		return
	}
	data := make([]interface{}, len(foods))
	for i := range foods {
		data[i] = contracts.FoodResponse{
			Id: foods[i].ID,
			Name: foods[i].Name,
			Description: foods[i].Description,
			Price: foods[i].Price,
			PathImg: foods[i].PathImg,
			FoodCategory: foods[i].FoodCategory,
			IsAvailable: foods[i].IsAvailable,
		}
	}
	utils.SendSuccessArray(ctx, "find-all-foods-by-menu-id", data)
}

func FoodServiceUpdate(ctx *gin.Context) {
	logger := config.NewLogger("FOOD UPDATE HANDLER")
	var dto contracts.FoodRequest
	if err := ctx.BindJSON(&dto); err != nil {
		utils.SendError(ctx, 400, err.Error())
		return
	}

	if err := validates.ValidateFoodRequest(&dto); err != nil {
		logger.Errf("validation error: %v", err)
		utils.SendError(ctx, 400, err.Error())
		return
	}

	foodIdStr := ctx.Param("id")
	foodId, err := uuid.Parse(foodIdStr)

	food := schemas.Food{
		Name:         dto.Name,
		Description:  dto.Description,
		Price:        dto.Price,
		PathImg:      "",
		FoodCategory: dto.FoodCategory,
		MenuId:       dto.MenuId,
	}

	err = repos.UpdateFood(foodId, &food, ctx)
	if err != nil {
		utils.SendError(ctx, http.StatusInternalServerError, "error updating food")
		return
	}
	utils.SendSuccess(ctx, "update-food")
}

func FoodServiceDelete(ctx *gin.Context) {
	foodIdStr := ctx.Param("id")
	foodId, err := uuid.Parse(foodIdStr)
	if err != nil {
		utils.SendError(ctx, http.StatusBadRequest, "invalid food id")
		return
	}
	err = repos.DeleteFood(foodId, ctx)
	if err != nil {
		utils.SendError(ctx, http.StatusInternalServerError, "error deleting food")
		return
	}
	utils.SendSuccess(ctx, "delete-food")
}

File: Client/kerokume/src/services/foodService.ts
import { api } from '../lib/api'
import type { ApiListResponse } from '../types/api'
import type { Food, FoodPayload } from '../types/food'

type FoodApi = {
  id?: string
  Id?: string
  ID?: string
  name?: string
  Name?: string
  description?: string
  Description?: string
  price?: number
  Price?: number
  foodCategory?: Food['foodCategory']
  FoodCategory?: Food['foodCategory']
  isAvailable?: boolean
  IsAvailable?: boolean
  pathImg?: string
  PathImg?: string
}

function normalizeFood(food: FoodApi): Food {
  return {
    id: food.id ?? food.Id ?? food.ID ?? '',
    name: food.name ?? food.Name ?? 'Sem nome',
    description: food.description ?? food.Description ?? '',
    price: food.price ?? food.Price ?? 0,
    foodCategory: food.foodCategory ?? food.FoodCategory ?? 'COMIDA',
    isAvailable: food.isAvailable ?? food.IsAvailable ?? true,
    pathImg: food.pathImg ?? food.PathImg,
  }
}

export async function getFoodsByMenu(menuId: string) {
  const response = await api.get<ApiListResponse<FoodApi>>(`/foods/${menuId}`, {
    params: { menuId },
  })
  const rawFoods = response.data.data ?? []
  return rawFoods.map(normalizeFood).filter((food) => food.id)
}

export async function createFood(payload: FoodPayload) {
  await api.post('/foods', payload, {
    params: { menuId: payload.menuId },
  })
}

export async function updateFood(foodId: string, payload: FoodPayload) {
  await api.put(`/foods/${foodId}`, payload, {
    params: { menuId: payload.menuId },
  })
}

export async function deleteFood(foodId: string, menuId: string) {
  await api.delete(`/foods/${foodId}`, {
    params: { menuId },
    data: { menuId },
  })
}
